pkg/apis/core/validation/validation.go

```diff
@@ -3720,10 +3758,75 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
 	if len(pod.Spec.EphemeralContainers) > 0 {
 		allErrs = append(allErrs, field.Forbidden(fldPath.Child("ephemeralContainers"), "cannot be set on create"))
 	}
+	allErrs = append(allErrs, validateSeccompAnnotationsAndFields(pod.ObjectMeta, &pod.Spec, field.NewPath(""))...)
 
 	return allErrs
 }
 
+// ValidateSeccompAnnotationsAndFields iterates through all containers and ensure that when both seccompProfile and seccomp annotations exist they match.
+func validateSeccompAnnotationsAndFields(objectMeta metav1.ObjectMeta, podSpec *core.PodSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if podSpec.SecurityContext != nil && podSpec.SecurityContext.SeccompProfile != nil {
+		// If both seccomp annotations and fields are specified, the values must match.
+		if annotation, found := objectMeta.Annotations[v1.SeccompPodAnnotationKey]; found {
+			fldPath = fldPath.Child("securityContext").Child("seccompProfile")
+			allErrs = append(allErrs, validateSeccompAnnotationsAndFieldsMatch(annotation, podSpec.SecurityContext.SeccompProfile, fldPath))
+		}
+	}
+
+	index := 0
+	fldPath = fldPath.Child("containers")
```

Comment:
VisitContainersWithPath visits containers of all types (initContainers, containers, ephemeralContainers), so a single field path and increasing index is not correct... 

I would suggest the following:
1. change the signature of VisitContainersWithPath to `VisitContainersWithPath(podSpec *api.PodSpec, specPath *field.Path, visitor ContainerVisitorWithPath)`
2. Update all other existing call sites to pass in `field.NewPath("spec")` as the spec path
3. Update this call site to pass in the `specPath` arg